capsules/src/si7021.rs

```diff
@@ -239,25 +239,29 @@ impl<'a, A: time::Alarm<'a>> i2c::I2CClient for SI7021<'a, A> {
 
 impl<'a, A: time::Alarm<'a>> kernel::hil::sensors::TemperatureDriver<'a> for SI7021<'a, A> {
     fn read_temperature(&self) -> kernel::ReturnCode {
-        self.buffer.take().map_or_else(
-            || {
-                if self.on_deck.get() != OnDeck::Nothing {
-                    ReturnCode::EBUSY
-                } else {
-                    self.on_deck.set(OnDeck::Temperature);
-                    ReturnCode::SUCCESS
-                }
-            },
-            |buffer| {
+        // This chip handles both humidity and temperature measurements. We can
+        // only start a new measurement if the chip is idle. If it isn't then we
+        // can put this request "on deck" and it will happen after the
+        // temperature measurement has finished.
+        if self.state.get() == State::Idle {
+            self.buffer.take().map_or(ReturnCode::EBUSY, |buffer| {
```

Comment:
Not a big deal, but maybe this should be `ReturnCode::FAIL`, as its a logic error for the buffer to be missing rather than something that could happen with a certain sequence of calls?